src/mongo_client.ts

```diff
@@ -564,44 +557,26 @@ export class MongoClient extends TypedEventEmitter<MongoClientEvents> {
     optionsOrOperation?: ClientSessionOptions | WithSessionCallback,
     callback?: WithSessionCallback
   ): Promise<void> {
-    let options: ClientSessionOptions = optionsOrOperation as ClientSessionOptions;
-    if (typeof optionsOrOperation === 'function') {
-      callback = optionsOrOperation as WithSessionCallback;
-      options = { owner: Symbol() };
-    }
+    const options = {
+      // Always define an owner
+      owner: Symbol(),
+      // If it's an object inherit the options
+      ...(typeof optionsOrOperation === 'object' ? optionsOrOperation : {})
+    };
+
+    const withSessionCallback =
+      typeof optionsOrOperation === 'function' ? optionsOrOperation : callback;
 
-    if (callback == null) {
+    if (withSessionCallback == null) {
       throw new MongoInvalidArgumentError('Missing required callback parameter');
     }
 
     const session = this.startSession(options);
     const Promise = PromiseProvider.get();
 
-    let cleanupHandler: CleanUpHandlerFunction = ((err, result, opts) => {
-      // prevent multiple calls to cleanupHandler
-      cleanupHandler = () => {
-        // TODO(NODE-3483)
-        throw new MongoRuntimeError('cleanupHandler was called too many times');
-      };
-
-      opts = Object.assign({ throw: true }, opts);
-      session.endSession();
-
-      if (err) {
-        if (opts.throw) throw err;
-        return Promise.reject(err);
-      }
-    }) as CleanUpHandlerFunction;
-
-    try {
-      const result = callback(session);
-      return Promise.resolve(result).then(
-        result => cleanupHandler(undefined, result, undefined),
-        err => cleanupHandler(err, null, { throw: true })
-      );
-    } catch (err) {
-      return cleanupHandler(err, null, { throw: false }) as Promise<void>;
-    }
+    return Promise.resolve()
+      .then(() => withSessionCallback(session))
```

Comment:
technically, because the result is returned, the declared return type is incorrect, because the type returned would be the type of the return type of `withSessionCallback` and not guaranteed to be `Promise<void>`

Reply:
fixed!